account_commission/models/account_move.py

```diff
@@ -0,0 +1,233 @@
+# Copyright 2020 Tecnativa - Manuel Calero
+# Copyright 2014-2022 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from lxml import etree
+
+from odoo import _, api, exceptions, fields, models
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    commission_total = fields.Float(
+        string="Commissions",
+        compute="_compute_commission_total",
+        store=True,
+    )
+    partner_agent_ids = fields.Many2many(
+        string="Agents",
+        comodel_name="res.partner",
+        compute="_compute_agents",
+        search="_search_agents",
+    )
+
+    @api.depends("partner_agent_ids", "invoice_line_ids.agent_ids.agent_id")
+    def _compute_agents(self):
+        for move in self:
+            move.partner_agent_ids = [
+                (6, 0, move.mapped("invoice_line_ids.agent_ids.agent_id").ids)
+            ]
+
+    @api.model
+    def _search_agents(self, operator, value):
+        ail_agents = self.env["account.invoice.line.agent"].search(
+            [("agent_id", operator, value)]
+        )
+        return [("id", "in", ail_agents.mapped("object_id.move_id").ids)]
+
+    @api.depends("line_ids.agent_ids.amount")
+    def _compute_commission_total(self):
+        for record in self:
+            record.commission_total = 0.0
+            for line in record.line_ids:
+                record.commission_total += sum(x.amount for x in line.agent_ids)
+
+    def action_post(self):
+        """Put settlements associated to the invoices in invoiced state."""
+        self.mapped("line_ids.settlement_id").write({"state": "invoiced"})
+        return super().action_post()
+
+    def button_draft(self):
+        """Check settled lines and put settlements associated to the invoices in
+        exception.
+        """
+        if any(self.mapped("invoice_line_ids.any_settled")):
+            raise exceptions.ValidationError(
+                _("You can't cancel an invoice with settled lines"),
+            )
+        self.mapped("line_ids.settlement_id").write({"state": "except_invoice"})
```

Comment:
```suggestion
        self.mapped("line_ids.settlement_id").write({"state": "invoiced"})
```
We should not do this here for button_draft() as it makes the settlement stuck at `except_invoce` state with no way to re-create a vendor bill for the settlement. IMO, no state update should be done on settlement when vendor bill is set back to draft. Setting `except_invoice` on the settlement should be done in button_cancel() as it did in 14.0. Instead, we may actually update the settlement to `invoiced` state in case canceled vendor bill is reset to draft.  Unlink case should still be handled to update the state of the settlement back to `settled`.

Reply:
OK, I have put it back this on cancel method, but added a way to reset a settlement in except state to draft for being re-invoiced.